rclcpp/test/rclcpp/test_subscription_publisher_count_api.cpp

```diff
@@ -23,21 +23,78 @@
 
 #include "test_msgs/msg/empty.hpp"
 
-using test_msgs::msg::Empty;
+namespace
+{
 
-struct TestParameters
+template<typename ... Ts>
+class NodeCreationPolicy
 {
-  rclcpp::NodeOptions node_options;
-  std::string description;
+public:
+  rclcpp::NodeOptions & node_options()
+  {
+    return options_;
+  }
+
+private:
+  rclcpp::NodeOptions options_;
 };
 
-std::ostream & operator<<(std::ostream & out, const TestParameters & params)
+template<typename T, typename ... Ts>
+class NodeCreationPolicy<T, Ts...>
 {
-  out << params.description;
-  return out;
-}
+public:
+  NodeCreationPolicy()
+  {
+    gather<T, Ts...>(options_);
+  }
 
-class TestSubscriptionPublisherCount : public ::testing::TestWithParam<TestParameters>
+  rclcpp::NodeOptions & node_options()
+  {
+    return options_;
+  }
+
+private:
+  template<typename U>
+  static rclcpp::NodeOptions &
+  gather(rclcpp::NodeOptions & options)
+  {
+    return U::gather(options);
+  }
+
+  template<typename U, typename V, typename ... Ws>
+  static rclcpp::NodeOptions &
+  gather(rclcpp::NodeOptions & options)
+  {
+    return gather<V, Ws...>(U::gather(options));
+  }
+
+  rclcpp::NodeOptions options_;
+};
+
+struct UseCustomContext
+{
+  static rclcpp::NodeOptions & gather(rclcpp::NodeOptions & options)
+  {
+    auto context = rclcpp::Context::make_shared();
+    context->init(0, nullptr);
+    return options.context(context);
+  }
+};
+
+struct PrintTestDescription
+{
+  template<typename T>
+  static std::string GetName(int)
+  {
+    return T::description;
+  }
+};
```

Comment:
there's some repetition with `test_publisher_subscription_count_api.cpp`.
I'm not sure if moving common logic to a header is worth it.

Reply:
Thought about it, seemed unworthy.